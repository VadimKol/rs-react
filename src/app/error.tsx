'use client';

import Image from 'next/image';
import { type ReactNode } from 'react';

import error_img from '@/assets/images/loader.png';
import { useTheme } from '@/hooks/useTheme';

import styles from './error.module.scss';

export default function ErrorPage({ error }: { error: Error & { digest?: string } }): ReactNode {
  const { theme } = useTheme();

  return (
    <main className={theme === 'dark' ? 'main' : 'main light'}>
      <div className={styles.error}>
        <p className={styles.name}>ERROR</p>
        <Image className={styles.error_img} src={error_img} alt="Rick and Morty error" width={500} height={500} />
        <p className={styles.desc}>{error?.message}</p>
        <p className={styles.desc}>Please refresh the page</p>
      </div>
    </main>
  );
}
