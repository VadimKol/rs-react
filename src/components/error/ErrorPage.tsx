import { type ReactNode, useEffect, useState } from 'react';
import { isRouteErrorResponse, useRouteError } from 'react-router-dom';

import error_img from '@/assets/images/loader.png';
import { useTheme } from '@/hooks/useTheme';

import styles from './styles.module.scss';

export function ErrorPage(): ReactNode {
  const error = useRouteError();
  const [errorMessage, setErrorMessage] = useState<string>();
  const { theme } = useTheme();

  useEffect(() => {
    if (error) {
      if (isRouteErrorResponse(error)) {
        setErrorMessage(error.statusText);
      } else if (error instanceof Error) {
        setErrorMessage(error.message);
      } else {
        setErrorMessage(String(error));
      }
    }
  }, [error]);

  return (
    <main className={theme === 'dark' ? 'main' : 'main light'}>
      <div className={styles.error}>
        <p className={styles.name}>ERROR</p>
        <img className={styles.error_img} src={error_img} alt="Rick and Morty error" width={500} height={500} />
        <p className={styles.desc}>{errorMessage}</p>
        <p className={styles.desc}>Please refresh the page</p>
      </div>
    </main>
  );
}
