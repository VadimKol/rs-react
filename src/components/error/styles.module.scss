.error {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  gap: 16px;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 20px;
  background: radial-gradient(circle at 50% 50%, var(--gray) 0%, var(--text-color) 100%);
}

.error_img {
  display: block;

  @media screen and (width <= 550px) {
    width: 100%;
    height: 100%;
  }
}

.desc,
.name {
  margin: 0;
}

.desc {
  font-size: 2rem;
  color: var(--background-color);

  @media screen and (width <= 550px) {
    font-size: 1rem;
  }
}

.name {
  font-family: var(--secondary-font);
  font-size: 4rem;
  letter-spacing: 5px;
  color: var(--secondary-color);
}
