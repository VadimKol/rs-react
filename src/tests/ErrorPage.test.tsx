import { render } from '@testing-library/react';

import { ErrorPage } from '@/components/error/ErrorPage';

describe('Error page', () => {
  it('renders correctly', () => {
    const { container } = render(<ErrorPage error={new Error('Some error')} />);

    expect(container).toMatchSnapshot();
  });
});
