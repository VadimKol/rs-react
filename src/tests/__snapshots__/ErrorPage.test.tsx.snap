// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error page > renders correctly 1`] = `
<div>
  <main
    class="main"
  >
    <div
      class="_error_38939b"
    >
      <p
        class="_name_38939b"
      >
        ERROR
      </p>
      <img
        alt="Rick and Morty error"
        class="_error_img_38939b"
        height="500"
        src="/src/assets/images/loader.png"
        width="500"
      />
      <p
        class="_desc_38939b"
      >
        Some error
      </p>
      <p
        class="_desc_38939b"
      >
        Please refresh the page
      </p>
    </div>
  </main>
</div>
`;
